pkg/odo/cli/component/create.go

```diff
@@ -677,23 +681,105 @@ func (co *CreateOptions) Validate() (err error) {
+// Currenty type git with a non github url is not supported
+func (co *CreateOptions) downloadProject() error {
+	devObj, err := devfile.Parse(co.devfileMetadata.devfilePath)
+	if err != nil {
+		return err
+	}
+	projects := devObj.Data.GetProjects()
+	nOfProjects := len(projects)
+	if nOfProjects == 0 {
+		return errors.Errorf("No project found in devfile component.")
+	}
+
+	if nOfProjects > 1 {
+		log.Warning("Only downloading first project from list.")
+	}
+
+	project := projects[0]
+
+	path, err := os.Getwd()
+	if err != nil {
+		return errors.Errorf("Could not get the current working directory: %s", err)
+	}
+
+	if project.ClonePath != nil && *project.ClonePath != "" {
+		clonePath := *project.ClonePath
+		if runtime.GOOS == "windows" {
+			clonePath = strings.Replace(clonePath, "\\", "/", -1)
+		}
+
+		path = filepath.Join(path, clonePath)
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			err = os.MkdirAll(path, os.FileMode(0755))
+			if err != nil {
+				return errors.Wrap(err, "Failed creating folder with path: "+path)
+			}
+		}
+	}
+
+	err = util.IsValidProjectDir(path, co.devfileMetadata.devfilePath)
+	if err != nil {
+		return err
+	}
+
+	var zipUrl string
+	switch project.Source.Type {
+	case "git":
+		if strings.Contains(project.Source.Location, "github.com") {
+			zipUrl, err = util.GetGitHubZipURL(project.Source.Location)
+			if err != nil {
+				return err
+			}
+		} else {
+			return errors.Errorf("Project type git with non github url not supported")
+		}
+	case "github":
+		zipUrl, err = util.GetGitHubZipURL(project.Source.Location)
+		if err != nil {
+			return err
+		}
+	case "zip":
+		zipUrl = project.Source.Location
+	default:
+		return errors.Errorf("Project type not supported")
+	}
+
+	err = util.GetAndExtractZip(zipUrl, path)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // Run has the logic to perform the required actions as part of command
 func (co *CreateOptions) Run() (err error) {
 	if experimental.IsExperimentalModeEnabled() {
 		// Download devfile.yaml file and create env.yaml file
 		if co.devfileMetadata.devfileSupport {
-			err := co.EnvSpecificInfo.SetConfiguration("create", envinfo.ComponentSettings{Name: co.devfileMetadata.componentName, Namespace: co.devfileMetadata.componentNamespace})
-			if err != nil {
-				return errors.Wrap(err, "Failed to create env.yaml for devfile component")
+			if co.devfileMetadata.devfilePath == "" && !util.CheckPathExists(DevfilePath) {
+				err := util.DownloadFile(co.devfileMetadata.devfileRegistry+co.devfileMetadata.devfileLink, DevfilePath)
+				if err != nil {
+					return errors.Wrap(err, "Faile to download devfile.yaml for devfile component")
+				}
+				co.devfileMetadata.devfilePath = DevfilePath
 			}
 
-			if !util.CheckPathExists(DevfilePath) {
-				err = util.DownloadFile(co.devfileMetadata.devfileRegistry+co.devfileMetadata.devfileLink, DevfilePath)
+			if co.devfileMetadata.downloadSource {
```

Comment:
When we're downloading the source + devfile into an empty directory, it doesn't seem to properly download the devfile:
```
Johns-MacBook-Pro-3:testcreate johncollier$ odo create java-spring-boot --downloadSource true
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile compatibility [80921ns]
 ✓  Validating devfile component [210766ns]
 ✗  open /Users/johncollier/testcreate/devfile.yaml: no such file or directory
```